rp2040-hal/src/pio.rs

```diff
@@ -0,0 +1,598 @@
+    pub fn with_program(mut self, p: &'a Program) -> Self {
+        self.instruction_origin = p.origin;
+        self.instructions(&p.instructions)
+            .wrap(p.wrap)
+            .side_set(p.side_set)
+    }
+
+    /// Set the instructions of the program.
+    pub fn instructions(mut self, instructions: &'a [u16]) -> Self {
+        self.instructions = instructions;
+        self
+    }
+
+    /// Set the wrap source and target.
+    ///
+    /// The program will automatically jump from the wrap bottom to the wrap top in 0 cycles.
+    pub fn wrap(mut self, wrap: Wrap) -> Self {
+        self.wrap = wrap;
+        self
+    }
+
+    /// Sets the pins asserted by `SET` instruction.
+    ///
+    /// The least-significant bit of `SET` instruction asserts the state of the pin indicated by `base`, the next bit
+    /// asserts the state of the next pin, and so on up to `count` pins. The pin numbers are considered modulo 32.
+    pub fn set_pins(mut self, base: u8, count: u8) -> Self {
+        assert!(count <= 5);
+        self.set_base = base;
+        self.set_count = count;
+        // self.in_base = base;
+        // self.out_base = base;
+        // self.out_count = count;
+        self
+    }
+
+    /// Sets the pins asserted by `OUT` instruction.
+    ///
+    /// The least-significant bit of `OUT` instruction asserts the state of the pin indicated by `base`, the next bit
+    /// asserts the state of the next pin, and so on up to `count` pins. The pin numbers are considered modulo 32.
+    pub fn out_pins(mut self, base: u8, count: u8) -> Self {
+        assert!(count <= 5);
+        self.out_base = base;
+        self.out_count = count;
+        self
+    }
+
+    pub fn in_pin_base(mut self, base: u8) -> Self {
+        self.in_base = base;
+        self
+    }
+
+    pub fn jmp_pin(mut self, pin: u8) -> Self {
+        self.jmp_pin = pin;
+        self
+    }
+
+    pub fn side_set_pin_base(mut self, base: u8) -> Self {
+        self.side_set_base = base;
+        self
+    }
+
+    /// Set the side-set status.
+    pub fn side_set(mut self, side_set: SideSet) -> Self {
+        self.side_set = side_set;
+        self
+    }
+
+    /// Set buffer sharing.
+    ///
+    /// See [`Buffers`] for more information.
+    pub fn buffers(mut self, buffers: Buffers) -> Self {
+        self.fifo_join = buffers;
+        self
+    }
+
+    /// Set the clock divisor.
+    ///
+    /// The is based on the sys_clk. Set 1 for full speed. A clock divisor of `n` will cause the state machine to run 1
+    /// cycle every `n` clock cycles. For small values of `n`, a fractional divisor may introduce unacceptable jitter.
+    pub fn clock_divisor(mut self, divisor: f32) -> Self {
+        self.clock_divisor = divisor;
+        self
+    }
+
+    /// Build the config and deploy it to a StateMachine.
+    pub fn build<P: Instance>(self, pio: &PIO<P>, sm: &StateMachine<P>) -> Result<(), BuildError> {
```

Comment:
If I understand correctly, this already executes the state machine. So I would propose to rename this function to `PioBuilder::execute`.